### PR TITLE
executor: fix `update ignore` into not exists partition (#21984)

### LINKED ISSUE #21965: update ignore into partition table shouldn't return no partition error
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

prepare
```sql
create table t (a int) partition by list (a) (partition p0 values in (0,1));
insert ignore into t values (1);
update ignore t set a=2 where a=1;
```

### 2. What did you expect to see? (Required)

`update` statement execute successfully.

### 3. What did you see instead (Required)

```sql
test> update ignore t set a=2 where a=1;
(1526, 'Table has no partition for value 2')
```

### 4. What is your TiDB version? (Required)

```sql
select tidb_version();
+-------------------------------------------------------------------+
| tidb_version()                                                    |
+-------------------------------------------------------------------+
| Release Version: v4.0.0-beta.2-1868-g1a9852f85                    |
| Edition: Community                                                |
| Git Commit Hash: 1a9852f85f7b90b013cafdd136260367203a29c0         |
| Git Branch: master                                                |
| UTC Build Time: 2020-12-21 03:07:25                               |
| GoVersion: go1.14.4                                               |
| Race Enabled: false                                               |
| TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306 |
| Check Table Before Drop: false                                    |
+-------------------------------------------------------------------+
```

relate issue : https://github.com/pingcap/tidb/issues/21894

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #21984 to release-4.0

---

### What problem does this PR solve?

Issue Number: close #21965 

Problem Summary:

### What is changed and how it works?

What's Changed:
fix `update ignore` into not exists partition
How it Works:
ignore rows that cannot found partition

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- No release note

### CODE REVIEW

File: executor/executor.go [hunk 1]
```diff
@@ -1773,6 +1773,22 @@ func ResetContextOfStmt(ctx sessionctx.Context, s ast.StmtNode) (err error) {
 	return
 }
 
+<<<<<<< HEAD
```
Comment (qw4990) [inline executor/executor.go:1776]: PTAL @TszKitLo40 @crazycs520

**Comment by crazycs520** [comment]
@lysu @jackysp PTAL

File: executor/write.go [hunk 2]
```diff
@@ -177,6 +180,10 @@ func updateRecord(ctx context.Context, sctx sessionctx.Context, h int64, oldData
 			// If the new handle exists, this will avoid to remove the record.
 			err = tables.CheckHandleExists(ctx, sctx, t, newHandle, newData)
 			if err != nil {
+				if terr, ok := errors.Cause(err).(*terror.Error); sctx.GetSessionVars().StmtCtx.IgnoreNoPartition && ok && terr.Code() == errno.ErrNoPartitionForGivenValue {
+					//return false, nil
```
Comment (tiancaiamao) [inline executor/write.go:184]: clean code

**Comment by tiancaiamao** [comment]
LGTM

**Comment by lysu** [comment]
LGTM

**Review by qw4990 — APPROVED** [review]

**Pull request merged**